# 3.5.3 ソフトマックス関数の特徴

In [14]:
import numpy as np

In [18]:
def softmax(a):
    c = np.max(a)
    exp_a = np.exp(a - c)
    sum_exp_a = np.sum(exp_a)
    y = exp_a / sum_exp_a
    return y

In [19]:
a = np.array([0.3, 2.9, 4.0])

In [20]:
y = softmax(a)

In [21]:
print(y)

[ 0.01821127  0.24519181  0.73659691]


In [22]:
np.sum(y)

1.0

- ソフトマックス関数の出力の総和は１となる。
- この性質のおかげで、ソフトマックス関数の出力は「確率」として解釈できる
- 上の例では以下のように解釈できる
  * y[0]の確率は 0.018(1.8%)
  * y[1]の確率は 0.245(24.5%)
  * y[2]の確率は 0.737(73.7%)


In [23]:
np.sum(a)

7.1999999999999993

In [24]:
a[0] / np.sum(a)

0.041666666666666671

In [25]:
np.sum(a) / np.sum(a)

1.0